### PR TITLE
Use bind params for basic where strings

### PR DESCRIPTION
**Pull request by composerinteralia**
This is something of a work in progress that starts to address #31219.

`Post.where('id = ?', 1)` used to generate this query:

```sql
 SELECT  "posts".* FROM "posts" WHERE ("id" = 1) LIMIT ?  [["LIMIT", 11]]
```

now it will generate this query with a bind parameter:

```sql
SELECT  "posts".* FROM "posts" WHERE "posts"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 11]]
```

I did the same thing for some other simple cases:

```ruby
Post.where('id != ?', 1)
Post.where('id <> ?', 1)
Post.where('id <= ?', 1)
Post.where('id >= ?', 1)
Post.where('id < ?', 1)
Post.where('id > ?', 1)
```

It also works with associations:

```ruby
Post.joins(:comments).where('comments.id = ?', 1)
Comment.joins(:post).where('comments.id = ?', 1)
```

And we can use `includes` without explicitly calling `references`:

```ruby
Post.includes(:comments).where('comments.id = ?', 1)
```

There are cases that this does NOT cover, and I am unsure how many of them we should try to cover:
- Keyword binds: `where('id = :id', id: 1) - this one seems doable,
- Quoted table and column names: `where('"posts"."id" = ?', 1)- not sure if it makes sense to deal with this
- Multiple conditions with `AND` or `OR` - Maybe not? This might get out of hand
- Expressions with "LIKE", "IN", "SIMILAR TO"

I'm guessing ideal solution is somewhere between what I have now and a library that parses SQL where clauses into ASTs 😄 .

### CODE REVIEW

File: activerecord/lib/active_record/relation/where_clause_factory.rb [hunk 1]
```diff
@@ -11,25 +127,36 @@ def initialize(klass, predicate_builder)
...
+          clause = HashClause.new(opts, @klass, @predicate_builder)
         when Arel::Nodes::Node
-          parts = [opts]
+          clause = ArelClause.new(opts)
         else
           raise ArgumentError, "Unsupported argument type: #{opts} (#{opts.class})"
         end
 
-        WhereClause.new(parts)
+        [WhereClause.new(clause.predicates), clause.references]
```
Comment (composerinteralia) [inline activerecord/lib/active_record/relation/where_clause_factory.rb]: Probably not ideal to return an array here. Could just return `clause` and build the `WhereClause` in `query_methods`, but then this class wouldn't be called `WhereClauseFactory` anymore. Suggestions are most welcome!

File: activerecord/lib/active_record/relation/where_clause_factory.rb [hunk 2]
```diff
@@ -11,25 +127,36 @@ def initialize(klass, predicate_builder)
...
+
+        def string_or_array_clause(all_opts)
+          if match = bind_string_match(all_opts)
+            BindStringClause.new(match[1], all_opts[1], match[2], @predicate_builder)
+          else
+            ArrayClause.new(all_opts, @klass)
+          end
+        end
+
+        BIND_STRING = /^([a-zA-Z_][a-zA-Z0-9_\.]*)(<=|>=|<>|!=|=|<|>)\?$/
```
Comment (composerinteralia) [inline activerecord/lib/active_record/relation/where_clause_factory.rb]: Will probably need to ditch the regex if we want to handle any of the more complicated cases

File: activerecord/test/cases/relation_test.rb [hunk 3]
```diff
@@ -180,14 +180,14 @@ def test_references_values_dont_duplicate
     test "merging a hash interpolates conditions" do
       klass = Class.new(FakeKlass) do
         def self.sanitize_sql(args)
-          raise unless args == ["foo = ?", "bar"]
-          "foo = bar"
+          raise unless args == ["foo = ? AND 1 = 1", "bar"]
```
Comment (composerinteralia) [inline activerecord/test/cases/relation_test.rb]: I will need to go through all of the existing tests that use where strings with binds to make sure they are doing what they are supposed to. `AND 1 = 1` was just a quick fix to get this test passing. I will come up with something better.

File: activerecord/lib/active_record/relation/where_clause_factory.rb [hunk 4]
```diff
@@ -11,25 +127,35 @@ def initialize(klass, predicate_builder)
...
+        def string_or_array_clause(all_opts)
+          if match = bind_string_match(all_opts)
+            BindStringClause.new(match[1], all_opts[1], match[2], @predicate_builder)
+          else
+            ArrayClause.new(all_opts, @klass)
+          end
+        end
 
-        attr_reader :klass, :predicate_builder
+        BIND_STRING = /^([a-zA-Z_][a-zA-Z0-9_\.]*)(<=|>=|<>|!=|=|<|>)\?$/
```
Comment (composerinteralia) [inline activerecord/lib/active_record/relation/where_clause_factory.rb]: Will probably need to ditch the regex if we want to handle any of the more complicated cases

**Comment by composerinteralia** [comment]
r? @sgrif 
I know you have thought about this a bit

**Comment by matthewd** [comment]
This is probably a dumb question, but why do we need to parse this much? I was hoping we could pass all placeholders straight through as bind parameters without caring how they're being used at all. (And deriving the type, where necessary, from the supplied value.)

**Comment by composerinteralia** [comment]
I don't think that is a dumb question at all. We definitely don't want to make changes like this if there isn't a good reason. 

My understanding is that it has to do with prepared statements.

For the query `Post.where("id = ?", 1)`:

If we pass `["id = ?", 1]` to `sanitize_sql` (as we currently do), we get back the string `"id = 1"`. With that we end up making a query that cannot be prepared:

```sql
 SELECT  "posts".* FROM "posts" WHERE (id = 1) LIMIT ?  [["LIMIT", 11]]
```

If we do some parsing we can skip `sanitize_sql` and just make the relevant Arel node with an `Arel::Nodes::BindParam`. Then we end up with a preparable query:

```sql
SELECT  "posts".* FROM "posts" WHERE "posts"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 11]]
```

This should also give more context: https://github.com/rails/rails/commit/cbcdecd2c55fca9613722779231de2d8dd67ad02

I also think it would be good to do some benchmarking on this.

**Comment by matthewd** [comment]
Right; my question is whether we can convert `["id = ?", 1]` directly into something like `Arel::Nodes::LiteralSequence.new([Arel::Nodes::SqlLiteral.new("id = "), Arel::Nodes::BindParam.new(ActiveRecord::Relation::QueryAttribute.new("_xx", 1, ActiveModel::Type::Integer.new))])`, without needing to specifically recognise the contained operation(s) or columns.

**Comment by composerinteralia** [comment]
Got it. I have no idea, but now I have something to go dig into. Thanks!

**Comment by matthewd** [comment]
:+1: (note that I invented `LiteralSequence`)

**Comment by composerinteralia** [comment]
@matthewd that suggestion is working great so far, and I'm really happy to be learning more about arel.

I do have a question about whether it makes sense to turn basically every string passed to `where` into one of these `LiteralSequence`s, or if we want to limit it to some subset. It would be fairly simple at this point to split on `?` and turn them all into `BindParam`s (although there are still some cases I am unsure about like `where('id IN (?)', [1,2,3])`). But I worry about the prepared statement cache filling up if all these queries become preparable.

A related problem, we could end up with multiple versions of essentially the same query in the cache:

```rb
Post.where(id: 1)
#=> SELECT  "posts".* FROM "posts" WHERE "posts"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 11]]

Post.where('id = ?' 1)
#=> SELECT  "posts".* FROM "posts" WHERE id = ? LIMIT ?  [["id", 1], ["LIMIT", 11]]

Post.where('posts.id = ?' 1)
#=> SELECT  "posts".* FROM "posts" WHERE posts.id = ? LIMIT ?  [["id", 1], ["LIMIT", 11]]
```

One thing that could help is modify `ConnectionAdapters::StatementPool` to eject by some combination of frequency and age, rather than rejecting the key inserted longest ago. I would be happy to add that if it is something we are interested in.

**Comment by matthewd** [comment]
> we could end up with multiple versions of essentially the same query in the cache

That part's fine; it's strictly better than we have now (under at least some conditions, where an embedded-literal-carrying expression ends up in a cacheable query), and is [approximately, under sane application behaviour] bounded to the number of call sites.

> every string passed to where

This may ultimately nudge up against #27947.. though `where` is somewhat unique among https://rails-sqli.org/ because its argument is always a condition-expression of some form: no-one could be passing a user-supplied string under the belief it can only reference an existing column, for example. (Speaking of that PR, see https://github.com/rails/rails/pull/27947#issuecomment-335977614 for a half-thought that ties in here.)

> `where('id IN (?)', [1,2,3])`

This is fine to do the simple thing (i.e., treat it identically to `where('id IN (?,?,?)', 1, 2, 3)`). I'd ultimately like to do other, smarter things (`id = ANY($1::int[])` in PostgreSQL, for example)... but it's all about building in layers, and `where(id: [1,2,3])` already produces the three-bind version. Changing that one day would be nice, but it can be separate from this goal.

> modify `ConnectionAdapters::StatementPool` to eject by some combination of frequency and age

I don't think we have any concrete idea of how big/full people's statement pools currently get. This does seem worth exploring, though I suspect a lot of people who are using PostgreSQL and whose applications are large enough to risk filling the pool, are likely using pgbouncer in transaction pooling mode, and therefore have prepared statements turned off entirely.

**Comment by matthewd** [comment]
> though `where` is somewhat unique among https://rails-sqli.org/ because its argument is always a condition-expression of some form: no-one could be passing a user-supplied string under the belief it can only reference an existing column, for example.

@mastahyeti do you have any thoughts on that point? I think the sensible goal is that basically everything else in https://rails-sqli.org/ should transition to requiring `Arel.sql` as `order` now does... but I'm feeling a bit hesitant about `where`: there's nothing the caller could've meant to provide *other than* an SQL fragment.

**Comment by mastahyeti** [comment]
>> though where is somewhat unique among https://rails-sqli.org/ because its argument is always a condition-expression of some form: no-one could be passing a user-supplied string under the belief it can only reference an existing column, for example.

> @mastahyeti do you have any thoughts on that point? I think the sensible goal is that basically everything else in https://rails-sqli.org/ should transition to requiring Arel.sql as order now does... but I'm feeling a bit hesitant about where: there's nothing the caller could've meant to provide other than an SQL fragment.

I think I agree. The main purpose of `#where` is t accept SQL. It might be good to think about cases like `Users.where(params[:column] => params[:value])` though?

**Comment by composerinteralia** [comment]
I didn't think there was an injection risk with the hash syntax. With `Users.where(params[:column] => params[:value])` those should both get quoted.

For strings we could consider string + other arguments as relatively safe,
since somebody using the bind arguments is less likely to interpolate bad things directly into the string.
```rb
User.where('id = ?', 1)
```
But passing a single string argument is unsafe and requires wrapping with Arel.sql():
```rb
User.where(Arel.sql("id = #{i_promise_this_is_not_user_input}"))
```

@matthewd Thoughts?

**Comment by matthewd** [comment]
I'm reluctant to assume a simple string contains dangerous interpolation, just because I'm reluctant to assume a simple string contains any interpolation at all. 

I think we'd be getting in the way of too many `'NOT active'`s and `publish_at <= now()`s, trying to defend against something that is well-known to be Bad Code 101. (And by that I don't just mean "that's so bad they deserve the risk / don't deserve protection", but rather "someone doing that will just as happily paste the `Arel.sql` call in there too": I'd love to save them, but I don't think this mechanism can help.)

I believe the hash form is safe within reason, particularly now that Parameters doesn't inherit from Hash: the key should always get quoted as a column name (which still seems a data-risky thing to expose to user input, but is clearly the author's intent), and the value will be safely handled as a query value (at worst, an array -- but barring a rather unusual query, that's unlikely to expose more than the client could've achieved with two separate requests containing single values).

**Comment by composerinteralia** [comment]
@matthewd I played around with this for a bit, but without great results. The current version looks like this: https://github.com/composerinteralia/rails/compare/master...composerinteralia:where-binds

The idea was to put shared logic related to bind variables in `SqlLiteral`, then use collectors to get either a string for `#santize_sql_array`, or a `LiteralSequence` for `#where`. This doesn't seem to have improved `#where` in any meaningful way, and it made `#sanitize_sql_array` quite a bit slower. If I leave `#sanitize_sql_array` alone, which I think I have to, I end up largely duplicating that method to generate the `LiteralSequence` for `#where`.

I'm happy to keep working on this if you think I should, but at the moment I do not have a good solution in mind.

**Comment by composerinteralia** [comment]
On another note, I remember seeing somewhere that we might want to move Arel into Rails. If that is indeed the case, I would be happy to work on it.

**Comment by roooodcastro** [comment]
Sorry to raise this PR from the dead, but has this been completely dropped or is it something that might be worked on in the future?

I'd also like to give my opinion on a possible implementation of this: instead of trying to parse the given string for `?` and column names, which can get messy, we could have something like this:

```ruby
Post.where('id > ?', { type: 'int', value: 1 })
Post.where('id > :id', id: { type: 'int', value: 1 })
```

Which would generate this:

```sql
SELECT  "posts".* FROM "posts" WHERE "posts"."id" > ? [["id", 1]]
```

The reason for the hash syntax is that, if I remember correctly, you need the value and the column type to assign a bind variable (on some DBMSs like MSSQL you also need to explicitly specify the type), so instead of parsing it and trying to extrapolate the column type, we have the user specify it for us. Of course, if the parameter is not a hash, the current implementation would be called. This might break queries where a hash is intentionally passed as a parameter, in which case a specialised class could be used instead of a simple hash.

This would be good not only to avoid parsing, but to also make it easier to identify arrays and use DB features such as ANY for PostgreSQL or SPLIT_STRING for MSSQL, for example.

**Pull request closed without merging**